OpenBSD pledge(2) and unveil(2) support.

vext01 opened the pull request:
Please don't merge this until another OpenBSD developer has checked my work.

This diff improves the security of Gophernicus when used on OpenBSD via `pledge(2)` and `unveil(2)`.

For more info on these APIs:
 * https://man.openbsd.org/pledge
 * https://man.openbsd.org/unveil

The only minor nit I have with my work is that we lose the ability to unveil if personal gopherspaces are enabled. The reason for this is that user homes are arbitrary paths, so simply unveiling `/home` wouldn't be right. We could add a switch which allows the user to specify which directories to unveil(2)... Not sure if it's worth it though and we already have so many switches.

Thanks

vext01 commented:
(I'd also like to squash those three commits into one before merge)

fosslinux commented:
I can squash wehn I murge.

jamestomasino commented:
pledge limits the types of work the program can do to a pretty common set. If that is exceeded (through poor code design or through privilege escalation attempts) the program is killed. This is nice!

unveil is like a mini chroot, defining paths that can be seen by the program. Gophernicus doesn't need to mess with certain paths and this help define that. Unfortunately, as @vext01 was saying, it's hard to name what all these paths are since userdir gophermaps could potentially be anywhere. If that's something that can be interpreted at runtime by checking for switches or falling back on the default path I think that would work.

vext01 commented:
So, should I perhaps add a switch for "extra unveil paths"? What do the gophernicus developers think?

fosslinux commented:
Yes, agreed, add a switch. On by default if userdirs are disabled, off if not.

vext01 commented:
> Yes, agreed, add a switch. On by default if userdirs are disabled, off if not.

I'm not sure how the switch can be "on by default". It'd need to be a list of extra paths to unveil. We could default it to `/home` if user dirs are on, but I'd rather not in case of misconfiguration.

I'm not sure if the current CLI parser allows for multiple appearances of the same switch, but if not, I'll make my new switch accept a colon or comma separated list of paths.

fosslinux commented:
Oh. What I was thinking was (logic path):

if your new flag is on {
    turn on your new stuff
}
else  if `-nu` specified on command line
{
    turn on your new stuff
}

jamestomasino commented:
If the setting requires additional parameters, maybe it's best configured in the conf file rather than through a command line switch?

vext01 commented:
So I guess I'm proposing something like a `-ud` (unveil dir) switch which is only allowed with `-nu` is not present. Ideally it'd be possible to pass this many times, e.g. `-ud  /home/vext01 -ud /home/josh -ud /export/home/rms`.

> maybe it's best configured in the conf file rather than through a command line switch?

There's a config file? I can't see any reference to one anywhere.

jamestomasino commented:
Huh, well I could have sworn there was one. I guess not. Just the defaults file in /etc/defaults

vext01 commented:
I've added a `-U` switch which is used to pass extra unveil paths.

Can someone look at #35 before I try to sync my work with master?

fosslinux commented on options.c:
```diff
@@ -128,7 +128,9 @@ void parse_args(state *st, int argc, char *argv[])
 			case 'L': sstrlcpy(st->server_location, optarg); break;
 			case 'A': sstrlcpy(st->server_admin, optarg); break;
-
+#ifdef __OpenBSD__
```
There should probably be some kind of check here, so that if it isn't obsd, it errors out.

vext01 replied:
Good point. I've made the addition of `-U` compile-time conditional.

fosslinux commented:
but otherwise looks great!

vext01 commented:
Fixed a bug, addressed @fosslinux's comment, updated docs, synced branch.

I hope we are converging now. There's two more things I'd like to do:
 * Get another OpenBSD dev to review the change (I've put out a toot on mastodon)
 * Squash the change before merge.

vext01 commented on options.c:
```diff
@@ -97,7 +97,11 @@ void parse_args(state *st, int argc, char *argv[])
-	while ((opt = getopt(argc, argv, "h:p:T:r:t:g:a:c:u:m:l:w:o:s:i:k:f:e:R:D:L:A:P:n:dbv?-")) != ERROR) {
+	while ((opt = getopt(argc, argv,
+#ifdef __OpenBSD__
+	    "U:" // unveil(2) paths are OpenBSD only.
```
I suppose that should be a `/* ... *.` style comment?

fosslinux commented:
@vext01 any progress?

vext01 commented:
One developer gave this a quick review, but admitted it wasn't a very thorough.

I'm awaiting one other developer. I will prod him now.

fosslinux commented:
Cool great

ibara commented:
@vext01 asked me to comment here so I'll just paste my previous comments:
The pledge stuff looks right. And works fine for me. I assume it works fine for you too. I'm much less knowledgeable of unveil but if it works for you then I say go for it since unveil might still shake out over time.

vext01 commented:
I've resolved the huge merge conflict caused by the tabs and spaces conversion.

I think I should re-test in case I fudged the merge. Please wait until I've tested before merging. I'd also like to squash.

vext01 commented:
Testing revealed one bug (personal spaces would not work without `-nm`). Fixed.

I was going to make all of the log messages I've added `LOG_DEBUG` instead of `LOG_INFO`, but I can't figure out how to turn on debug logging. Looks like it should be `-d` but it doesn't seem to work (at least not for me).

Any ideas?

vext01 commented:
ping? Any ideas on the above query?

fosslinux commented on gophernicus.c:
```diff
@@ -523,6 +531,91 @@ int main(int argc, char *argv[])
+         */
+        if (st.opt_personal_spaces) {
+            if (st.opt_syslog) {
+                syslog(LOG_INFO, "unveiling /etc/pwd.db");
```
debug

fosslinux commented on gophernicus.c:
```diff
@@ -523,6 +531,91 @@ int main(int argc, char *argv[])
+                continue; /* empty path */
+
+            if (st.opt_syslog) {
+                syslog(LOG_INFO,
```
debug

fosslinux commented on gophernicus.c:
```diff
@@ -523,6 +531,91 @@ int main(int argc, char *argv[])
+        /* Executable maps shell-out using popen(3) */
+        if (st.opt_exec) {
+            strlcat(pledges, " exec", sizeof(pledges));
+            syslog(LOG_INFO,
```
debug

fosslinux commented on gophernicus.c:
```diff
@@ -523,6 +531,91 @@ int main(int argc, char *argv[])
+        /* Personal spaces require getpwnam(3) and getpwent(3) */
+        if (st.opt_personal_spaces) {
+            strlcat(pledges, " getpw", sizeof(pledges));
+            syslog(LOG_INFO,
```
debug

fosslinux commented on gophernicus.c:
```diff
@@ -544,31 +637,32 @@ int main(int argc, char *argv[])
+            }
+        }
+    } else {
+        shm = NULL;
```
Good catch, didn't notice this before

fosslinux commented:
Sorry, I've been super busy IRL. 

> I was going to make all of the log messages I've added LOG_DEBUG instead of LOG_INFO, but I can't figure out how to turn on debug logging. Looks like it should be -d but it doesn't seem to work (at least not for me).

```
-d            Debug output in syslog and /server-status
```

In syslog? I'll look a bit further later today

vext01 commented:
> In syslog? I'll look a bit further later today

Yeah, seems broken.

Running `gophernicus -d` and fetching `/` before my changing any of my `INFO`s to `DEBUG`s:
```
Aug 31 15:05:50 arrakis gophernicus[69210]: executable gophermaps are enabled, no unveil(2)
Aug 31 15:05:50 arrakis gophernicus[69210]: shared-memory enabled, can't pledge(2)
Aug 31 15:05:51 arrakis gophernicus[69210]: client sent us "/"
Aug 31 15:05:51 arrakis gophernicus[69210]: path to resource is "/var/gopher/"
Aug 31 15:05:51 arrakis gophernicus[69210]: request for "gopher://arrakis.home:70/1/" from unknown
Aug 31 15:05:51 arrakis gophernicus[69210]: parsing static gophermap "/var/gopher//gophermap"
```

Now change my `INFO`s to `DEBUG`s:
```
Aug 31 15:09:15 arrakis gophernicus[33197]: client sent us "/"
Aug 31 15:09:15 arrakis gophernicus[33197]: path to resource is "/var/gopher/"
Aug 31 15:09:15 arrakis gophernicus[33197]: request for "gopher://arrakis.home:70/1/" from unknown
Aug 31 15:09:15 arrakis gophernicus[33197]: parsing static gophermap "/var/gopher//gophermap"
```

The debug messages are missing from `/var/log/daemon`.

I've opted to change *all* of the pledge and unveil messages to `DEBUG` as otherwise due to the way inetd works, we print that stuff on every hit, which will just junk up the logs IMHO.

I've added a note to the README saying that you can see what pledge and unveil are doing by using `-d`.

**But `-d` needs to be fixed before this can be merged.**

fosslinux commented:
Ahha.

fosslinux commented:
In options.c, `-d` makes `st->debug = TRUE`.

So looking for where `st->debug` is referenced:

```
➜ grep 'st->debug' -r .
./platform.c:    if (st->debug) {
./options.c:            case 'd': st->debug = TRUE; break;
./options.c:    if (!st->opt_syslog) st->debug = FALSE;
./menu.c:    if (st->debug) {
./file.c:    if (st->debug) syslog(LOG_INFO, "outputting binary file \"%s\"", st->req_realpath);
./file.c:    if (st->debug) syslog(LOG_INFO, "outputting binary file \"%s\"", st->req_realpath);
./file.c:    if (st->debug) syslog(LOG_INFO, "outputting text file \"%s\"", st->req_realpath);
./file.c:            if (st->debug) {
./file.c:    if (st->debug) syslog(LOG_INFO, "executing script \"%s\"", script);
./file.c:    else if (st->debug) syslog(LOG_INFO, "script \"%s\" was blocked by -nx", script);
./gophernicus.c:            if (st->debug) {
./gophernicus.c:    st->debug = FALSE;
```

So, actually it **dosen't enable LOG_DEBUG, it simply is used as an indicator of whether we are in debugging mode**.

So, it's not a bug, it's a feature ;)

Tbh, I don't think it's a great way to do it. I'll look into it but recommend you do it this way for now.

vext01 commented:
Here we go. Also removed unnecessary ` if (st.opt_syslog) {` checks. It seems `syslog(3)` is simply a no-op if logging isn't enabled, so that's fine.

**If you think this is now ready, please let me squash the commits before we merge.**

Thanks

fosslinux commented:
Yes, I'm ready for merge. you can squash or i can squash on merge

vext01 commented:
Squashed.

(Your tabs to spaces sweep made this quite hard. I'd recommend against such sweeping changes in the future. Especially while there are PRs open)

fosslinux commented:
Yeah, I realised that it'd cause problems... Just didn't expect it to have such an impact. Sorry
